From 9d4ea8bb75b2c56d4713fea42a7b5913468a215b Mon Sep 17 00:00:00 2001
From: Peter Harliman Liem <peter.harliman.liem@intel.com>
Date: Mon, 27 May 2019 11:38:31 +0800
Subject: [PATCH] DRVLIB_SW-2331 - Add prx300 gphy reset in shutdown

Global sw reset in prx300 will not reset RCU register
for gphy, which eventually causes issue at reboot.
We therefore explicitily reset the gphy during shutdown.
---
 drivers/net/ethernet/lantiq/xrx500_phy_fw.c | 26 ++++++++++++++++++++++++++
 1 file changed, 26 insertions(+)

diff --git a/drivers/net/ethernet/lantiq/xrx500_phy_fw.c b/drivers/net/ethernet/lantiq/xrx500_phy_fw.c
index 43ae44d40ffa..e6f5ec149de3 100644
--- a/drivers/net/ethernet/lantiq/xrx500_phy_fw.c
+++ b/drivers/net/ethernet/lantiq/xrx500_phy_fw.c
@@ -53,6 +53,7 @@ struct xway_gphy_data {
 	const struct xway_gphy_soc_data {
 		int (*boot_func)(struct xway_gphy_data *);
 		int (*dt_parse_func)(struct xway_gphy_data *);
+		int (*shutdown)(struct xway_gphy_data *);
 		int align;
 		int fixed_mem_size;
 	} *soc_data;
@@ -475,6 +476,20 @@ static int prx300_dt_parse(struct xway_gphy_data *priv)
 	return 0;
 }
 
+static int prx300_gphy_shutdown(struct xway_gphy_data *priv)
+{
+	struct prx300_reset_control *rst = &priv->rst.prx300;
+
+	dev_dbg(priv->dev, "Shutting down p31g gphy\n");
+
+	/* Global sw reset in prx300 will not reset RCU register
+	 * for gphy, which eventually causes issue at reboot.
+	 * We therefore explicitily reset the gphy during shutdown.
+	 */
+	reset_control_assert(rst->gphy);
+	return 0;
+}
+
 static int xway_gphy_load(struct xway_gphy_data *priv)
 {
 	const struct firmware *fw;
@@ -515,6 +530,15 @@ static int xway_gphy_load(struct xway_gphy_data *priv)
 	return 0;
 }
 
+static void xway_phy_shutdown(struct platform_device *pdev)
+{
+	struct xway_gphy_data *priv;
+
+	priv = platform_get_drvdata(pdev);
+	if (priv->soc_data->shutdown)
+		priv->soc_data->shutdown(priv);
+}
+
 static int xway_phy_fw_probe(struct platform_device *pdev)
 {
 	int ret = 0;
@@ -568,6 +592,7 @@ static struct xway_gphy_soc_data xrx500_gphy_data = {
 
 static struct xway_gphy_soc_data prx300_gphy_data = {
 	.boot_func = &prx300_gphy_boot,
+	.shutdown = &prx300_gphy_shutdown,
 	.dt_parse_func = &prx300_dt_parse,
 
 	/* To workaround GPHY memory corruption issue,
@@ -592,6 +617,7 @@ MODULE_DEVICE_TABLE(of, xway_phy_match);
 
 static struct platform_driver xway_phy_driver = {
 	.probe = xway_phy_fw_probe,
+	.shutdown = xway_phy_shutdown,
 	.driver = {
 		.name = "phy-xrx500",
 		.owner = THIS_MODULE,
